### DISCUSSION TITLE
Wayland

### DISCUSSION THREAD

**Discussion by ghost** [MoAlyousef/zfltk#17] [Ideas]
#### What

Do you have an idea how to support Wayland in build.zig? Maybe it's difficult to mirror what we have with cargo features.

#### Why

I run a pure Wayland desktop and don't have Xwayland installed.

**Comment by ghost** [comment]
I would repeat the question for Gl(u)Window feature, but didn't start a separate thread.

**Reply by MoAlyousef** [reply]
It should be possible via build options. std.Build.option() function. This would enable features via `zig build -Dwayland=true`. 

Something like:
```
b.option(bool, "wayland", "builds for wayland");
```

I’ll have to try it when I get the chance.

**Reply by MoAlyousef** [reply]
I've added an option zfltk-use-wayland.
So you could do `zig build -Dzfltk-use-wayland=true`

The other option is using Sdk.init_with_opts()
```zig
const sdk = Sdk.init_with_opts(b, .{ .use_wayland = true });
```

I've also added GlutWindow. Not all the methods were added yet.

**Reply by ghost** [reply]
Thanks! Is GlutWindow linking enabled unconditionally and up to the window creation code to choose whether it's used?

**Reply by ghost** [reply]
That was quick, highly appreciated!! While we're at it, what do you think of mirroring https://github.com/fltk-rs/fltk-rs/issues/1429?

**Reply by ghost** [reply]
> I've also added GlutWindow. Not all the methods were added yet.

Thanks. I'll try it again once the missing methods are there. For now, as expected, all you can get is a black window.

**Reply by ghost** [reply]
For the record, re-running the build (and using cmake cache) works around the dbus.h issue until cfltk is updated. I've seen that your fix has been merged by Manolo, but no rush. Thanks a lot.

**Comment by MoAlyousef** [comment]
> Thanks! Is GlutWindow linking enabled unconditionally and up to the window creation code to choose whether it's used?

Yes. Recent Zig versions don't support older Windows versions so OpenGL is assumed to always be present. In Rust it's not the situation since on Windows 7, OpenGL support is patchy, and Rust still supports Windows 7. On Linux, OpenGL support requires the installation of a library, so it's always possible, even without hardware acceleration support.

> what do you think of mirroring https://github.com/fltk-rs/fltk-rs/issues/1429?

Zfltk doesn't link libgtk on wayland already. fltk-rs used to link it because it was previously required by libdecor.

>  For now, as expected, all you can get is a black window.

I haven't tried it yet, but OpenGL drawing should be supported. The default is a black window, unless you setBox(.flat) and setColor() to something:
```zig
// modified simple example
const zfltk = @import("zfltk");
const app = zfltk.app;
const GlutWindow = zfltk.GlutWindow;
const Button = zfltk.Button;
const Box = zfltk.Box;
const Color = zfltk.enums.Color;

fn butCb(but: *Button(.normal), data: ?*anyopaque) void {
    var box = Box.fromRaw(data.?);

    box.setLabel("Hello World!");

    but.setColor(Color.fromName(.cyan));
}

pub fn main() !void {
    try app.init();
    app.setScheme(.gtk);

    var win = try GlutWindow.init(.{
        .w = 400,
        .h = 300,

        .label = "Hello",
    });

    win.setBox(.flat);
    win.setColor(Color.fromName(.background));
    win.begin();

    var but = try Button(.normal).init(.{
        .x = 160,
        .y = 220,
        .w = 80,
        .h = 40,

        .label = "Click me!",
    });

    but.setDownBox(.flat);

    var box = try Box.init(.{
        .x = 10,
        .y = 10,
        .w = 380,
        .h = 180,

        .boxtype = .up,
    });

    box.setLabelFont(.courier);
    box.setLabelSize(18);

    win.end();
    win.show();

    but.setCallbackEx(butCb, box);
    try app.run();
}
```

**Reply by ghost** [reply]
I had forgotten to call `win.begin()` :facepalm:. It builds and runs correctly.

**Reply by ghost** [reply]
> On Linux, OpenGL support requires the installation of a library, so it's always possible, even without hardware acceleration support.

Specifically `libGLU`. Despite being called `GlutWindow`, it doesn't require GLUT, or is maybe optional.

**Reply by ghost** [reply]
I'm asking a lot without contributing code, so if you'll allow me, I want to put this idea out there: what do you think of exposing `--disable-x11` (in fltk-rs and zfltk)?

**Comment by MoAlyousef** [comment]
> Specifically libGLU. Despite being called `GlutWindow', it doesn't require GLUT, or is maybe optional.

It's optional.

> what do you think of exposing --disable-x11 (in cfltk and therefore fltk-rs plus zfltk)?

It's supported in fltk-rs via setting an env variable CFLTK_WAYLAND_ONLY which will basically set FLTK's OPTION_WAYLAND_ONLY=OFF during the build. Zfltk doesn't have such an option yet.

**Reply by ghost** [reply]
I'll give CFLTK_WAYLAND_ONLY a try

**Reply by ghost** [reply]
I can't tell if it worked as expected.

1. Switched fltk dependency to git
2. Created `.cargo/config.toml` with the following in it
```toml
[env]
CFLTK_WAYLAND_ONLY="ON"
```
3. Disabled `fltk-theme` as likely source of dependency solver mismatch, resulting in this Cargo.toml:
```toml
# [dependencies]
# fltk-theme = "0.7"

[dependencies.fltk]
# version = "1.4"
git = "https://github.com/fltk-rs/fltk-rs"
features = ["use-ninja", "use-wayland"]

[features]
default = ["use-glwin"]
use-glwin = ["fltk/enable-glwindow"]
```

Output of ldd:
```
	/lib64/ld-linux-x86-64.so.2 => /usr/lib64/ld-linux-x86-64.so.2 (0x0000723829a46000)
	libblkid.so.1 => /usr/lib/libblkid.so.1 (0x000072382809b000)
	libbrotlicommon.so.1 => /usr/lib/libbrotlicommon.so.1 (0x0000723828078000)
	libbrotlidec.so.1 => /usr/lib/libbrotlidec.so.1 (0x00007238280ff000)
	libbz2.so.1.0 => /usr/lib/libbz2.so.1.0 (0x000072382810d000)
	libcairo.so.2 => /usr/lib/libcairo.so.2 (0x000072382936f000)
	libcap.so.2 => /usr/lib/libcap.so.2 (0x00007238284eb000)
	libc.so.6 => /usr/lib/libc.so.6 (0x0000723829061000)
	libdatrie.so.1 => /usr/lib/libdatrie.so.1 (0x0000723828170000)
	libdbus-1.so.3 => /usr/lib/libdbus-1.so.3 (0x00007238296ca000)
	libEGL.so.1 => /usr/lib/libEGL.so.1 (0x000072382985c000)
	libexpat.so.1 => /usr/lib/libexpat.so.1 (0x0000723828120000)
	libffi.so.8 => /usr/lib/libffi.so.8 (0x0000723828f41000)
	libfontconfig.so.1 => /usr/lib/libfontconfig.so.1 (0x0000723828672000)
	libfreetype.so.6 => /usr/lib/libfreetype.so.6 (0x00007238285a4000)
	libfribidi.so.0 => /usr/lib/libfribidi.so.0 (0x0000723828ed0000)
	libgcc_s.so.1 => /usr/lib/libgcc_s.so.1 (0x000072382933a000)
	libgcrypt.so.20 => /usr/lib/libgcrypt.so.20 (0x00007238283a3000)
	libgio-2.0.so.0 => /usr/lib/libgio-2.0.so.0 (0x00007238287f7000)
	libGLdispatch.so.0 => /usr/lib/libGLdispatch.so.0 (0x0000723828fa9000)
	libglib-2.0.so.0 => /usr/lib/libglib-2.0.so.0 (0x00007238289cd000)
	libGL.so.1 => /usr/lib/libGL.so.1 (0x00007238297d6000)
	libGLU.so.1 => /usr/lib/libGLU.so.1 (0x0000723829780000)
	libGLX.so.0 => /usr/lib/libGLX.so.0 (0x0000723828f77000)
	libgmodule-2.0.so.0 => /usr/lib/libgmodule-2.0.so.0 (0x00007238281ca000)
	libgobject-2.0.so.0 => /usr/lib/libgobject-2.0.so.0 (0x00007238294a1000)
	libgpg-error.so.0 => /usr/lib/libgpg-error.so.0 (0x00007238280d9000)
	libgraphite2.so.3 => /usr/lib/libgraphite2.so.3 (0x000072382814b000)
	libharfbuzz.so.0 => /usr/lib/libharfbuzz.so.0 (0x00007238286fb000)
	liblz4.so.1 => /usr/lib/liblz4.so.1 (0x000072382827b000)
	liblzma.so.5 => /usr/lib/liblzma.so.5 (0x0000723828370000)
	libmount.so.1 => /usr/lib/libmount.so.1 (0x0000723828179000)
	libm.so.6 => /usr/lib/libm.so.6 (0x000072382924d000)
	libOpenGL.so.0 => /usr/lib/libOpenGL.so.0 (0x0000723828f4c000)
	libpango-1.0.so.0 => /usr/lib/libpango-1.0.so.0 (0x0000723829503000)
	libpangocairo-1.0.so.0 => /usr/lib/libpangocairo-1.0.so.0 (0x000072382935f000)
	libpangoft2-1.0.so.0 => /usr/lib/libpangoft2-1.0.so.0 (0x0000723828e7d000)
	libpcre2-8.so.0 => /usr/lib/libpcre2-8.so.0 (0x00007238281d1000)
	libpixman-1.so.0 => /usr/lib/libpixman-1.so.0 (0x00007238284f7000)
	libpng16.so.16 => /usr/lib/libpng16.so.16 (0x00007238286c1000)
	libstdc++.so.6 => /usr/lib/libstdc++.so.6 (0x0000723828c00000)
	libsystemd.so.0 => /usr/lib/libsystemd.so.0 (0x0000723828b18000)
	libthai.so.0 => /usr/lib/libthai.so.0 (0x0000723828ec5000)
	libwayland-client.so.0 => /usr/lib/libwayland-client.so.0 (0x000072382976e000)
	libwayland-cursor.so.0 => /usr/lib/libwayland-cursor.so.0 (0x0000723829764000)
	libwayland-egl.so.1 => /usr/lib/libwayland-egl.so.1 (0x000072382986e000)
	libX11.so.6 => /usr/lib/libX11.so.6 (0x0000723829587000)
	libXau.so.6 => /usr/lib/libXau.so.6 (0x0000723828276000)
	libxcb-render.so.0 => /usr/lib/libxcb-render.so.0 (0x0000723828e9a000)
	libxcb-shm.so.0 => /usr/lib/libxcb-shm.so.0 (0x0000723828e95000)
	libxcb.so.1 => /usr/lib/libxcb.so.1 (0x0000723828f14000)
	libXcursor.so.1 => /usr/lib/libXcursor.so.1 (0x0000723829576000)
	libXdmcp.so.6 => /usr/lib/libXdmcp.so.6 (0x000072382826c000)
	libXext.so.6 => /usr/lib/libXext.so.6 (0x0000723828eff000)
	libXfixes.so.3 => /usr/lib/libXfixes.so.3 (0x000072382956e000)
	libXinerama.so.1 => /usr/lib/libXinerama.so.1 (0x0000723829582000)
	libxkbcommon.so.0 => /usr/lib/libxkbcommon.so.0 (0x000072382971b000)
	libXrender.so.1 => /usr/lib/libXrender.so.1 (0x0000723828ef2000)
	libz.so.1 => /usr/lib/libz.so.1 (0x0000723828eab000)
	libzstd.so.1 => /usr/lib/libzstd.so.1 (0x000072382829d000)
	linux-vdso.so.1 (0x00007ffff03d9000)
```

**Reply by MoAlyousef** [reply]
Try:
```
CFLTK_WAYLAND_ONLY="1"
```

**Reply by MoAlyousef** [reply]
Normally a clean rebuild shouldn’t be needed since fltk-sys’s build script contains
```rust
    println!("cargo:rerun-if-env-changed=CFLTK_WAYLAND_ONLY");
```

**Reply by ghost** [reply]
Nice, good to know. Had already started a clean build and therefore deleted my quick comment.

**Reply by ghost** [reply]
I suppose it made a difference. Not quite what I expected, with enough X libs still linked, but certainly different to without `CFLTK_WAYLAND_ONLY="1"`

```diff
--- 1	2023-07-09 13:59:18.196603959 +0000
+++ 2	2023-07-09 13:59:30.163424275 +0000
@@ -49,11 +49,8 @@
 	libxcb-render.so.0 => /usr/lib/libxcb-render.so.0
 	libxcb-shm.so.0 => /usr/lib/libxcb-shm.so.0
 	libxcb.so.1 => /usr/lib/libxcb.so.1
-	libXcursor.so.1 => /usr/lib/libXcursor.so.1
 	libXdmcp.so.6 => /usr/lib/libXdmcp.so.6
 	libXext.so.6 => /usr/lib/libXext.so.6
-	libXfixes.so.3 => /usr/lib/libXfixes.so.3
-	libXinerama.so.1 => /usr/lib/libXinerama.so.1
 	libxkbcommon.so.0 => /usr/lib/libxkbcommon.so.0
 	libXrender.so.1 => /usr/lib/libXrender.so.1
 	libz.so.1 => /usr/lib/libz.so.1
```

**Reply by MoAlyousef** [reply]
It will depend on the platform. Most likey these are pulled by cairo and pango I think. If you try on a platform without x11, like a github actions runner with only cairo, pango, wayland libs, dbus and xkbcommon, you should get a cleaner list.

**Reply by MoAlyousef** [reply]
https://github.com/MoAlyousef/test_wayland/blob/main/.github/workflows/rust.yml

**Reply by ghost** [reply]
Direct dependencies extracted with `readelf -d`, therefore safe to say it works:
```
 0x0000000000000001 (NEEDED)             Shared library: [libwayland-egl.so.1]
 0x0000000000000001 (NEEDED)             Shared library: [libEGL.so.1]
 0x0000000000000001 (NEEDED)             Shared library: [libGL.so.1]
 0x0000000000000001 (NEEDED)             Shared library: [libGLU.so.1]
 0x0000000000000001 (NEEDED)             Shared library: [libwayland-client.so.0]
 0x0000000000000001 (NEEDED)             Shared library: [libwayland-cursor.so.0]
 0x0000000000000001 (NEEDED)             Shared library: [libxkbcommon.so.0]
 0x0000000000000001 (NEEDED)             Shared library: [libdbus-1.so.3]
 0x0000000000000001 (NEEDED)             Shared library: [libpango-1.0.so.0]
 0x0000000000000001 (NEEDED)             Shared library: [libgobject-2.0.so.0]
 0x0000000000000001 (NEEDED)             Shared library: [libcairo.so.2]
 0x0000000000000001 (NEEDED)             Shared library: [libpangocairo-1.0.so.0]
 0x0000000000000001 (NEEDED)             Shared library: [libgcc_s.so.1]
 0x0000000000000001 (NEEDED)             Shared library: [libm.so.6]
 0x0000000000000001 (NEEDED)             Shared library: [libc.so.6]
 0x0000000000000001 (NEEDED)             Shared library: [ld-linux-x86-64.so.2]
```

Unless I use Gentoo USE flags or an equivalent build environment without X, one of pango, pangocairo will expectedly pull in those indirect dependencies. It's fine, since nobody wants to have a pure Wayland distro yet, and it shouldn't have a runtime effect.

**Reply by ghost** [reply]
`lddtree` Python script from `pax-utils` confirms it's pulled in by cairo. It's unimportant, but I'm thinking out loud: would you expect a GlutWindow-only FLTK build (not what we have) to drop all Cairo use, or is it maybe used in shared code paths?

**Reply by ghost** [reply]
lddtree output:

```
    libcairo.so.2 => /usr/lib/libcairo.so.2
        libfontconfig.so.1 => /usr/lib/libfontconfig.so.1
            libexpat.so.1 => /usr/lib/libexpat.so.1
        libXext.so.6 => /usr/lib/libXext.so.6
        libXrender.so.1 => /usr/lib/libXrender.so.1
        libxcb-render.so.0 => /usr/lib/libxcb-render.so.0
        libxcb-shm.so.0 => /usr/lib/libxcb-shm.so.0
        libpixman-1.so.0 => /usr/lib/libpixman-1.so.0
```

**Reply by MoAlyousef** [reply]
> would you expect a GlutWindow-only FLTK build (not what we have) to drop all Cairo use, or is it maybe used in shared code paths?

It’s possible for x11 via enable-glwindow,no-pango. 

For wayland, cairo is used:
- for decorations via libdecor. 
- for font rasterization via pangocairo.

**Reply by ghost** [reply]
Out of curiosity, not that it makes sense in production code, I gave `no-pango` a try and saw link errors for missing cairo symbols:

```
ld.lld: error: undefined symbol: cairo_clip
    >>> referenced by Fl_Wayland_Graphics_Driver.cxx
    >>> <long line>
    >>> referenced by Fl_Cairo_Graphics_Driver.cxx
...
ld.lld: error: undefined symbol: cairo_get_target
    >>> referenced by Fl_Wayland_Graphics_Driver.cxx
    >>> <long line>
    >>> referenced by Fl_Wayland_Graphics_Driver.cxx
    >>> <long line>
    >>> referenced by fl_wayland_clipboard_dnd.cxx
```

**Reply by MoAlyousef** [reply]
Yeah no-pango won’t work on use-wayland. Actually no-pango is a bit of a misnomer, it should probably be renamed to no-pango-cairo or something like that.
